Tests/Runner/StateFileTest.php

```diff
@@ -142,79 +168,124 @@ public function testChange()
+                    self::assertArrayHasKey(StateFile::NAMESPACE_PREFIX, $data);
+                    self::assertArrayHasKey('key', $data[StateFile::NAMESPACE_PREFIX]);
+                    self::assertArrayHasKey('foo', $data[StateFile::NAMESPACE_PREFIX]);
+                    self::assertEquals('fooBar', $data[StateFile::NAMESPACE_PREFIX]['key']);
+                    self::assertEquals('bar', $data[StateFile::NAMESPACE_PREFIX]['foo']);
+                    return true;
+                })
+            );
         /** @var Client $sapiStub */
         $stateFile = new StateFile(
             $this->dataDir,
             $sapiStub,
             $this->encryptorFactory,
-            ['state' => 'fooBarBaz'],
+            ['key' => 'fooBarBaz'],
             'json',
             'docker-demo',
             'config-id',
-            new NullFilter()
+            new NullFilter(),
+            'row-id'
         );
-        $stateFile->stashState(["state" => "fooBar", "#foo" => "bar"]);
+        $stateFile->stashState(["key" => "fooBar", "foo" => "bar"]);
+        $stateFile->persistState();
     }
 
-    public function testChangeFromEmpty()
+    public function testPersistStateEncrypts()
     {
         $sapiStub = self::getMockBuilder(Client::class)
             ->disableOriginalConstructor()
             ->getMock();
         $sapiStub->expects(self::once())
             ->method('apiPut')
             ->with(
-                self::equalTo('storage/components/docker-demo/configs/config-id'),
-                self::equalTo(['state' => '{"state":"fooBar"}'])
+                $this->equalTo("storage/components/docker-demo/configs/config-id"),
+                $this->callback(function ($argument) {
+                    self::assertArrayHasKey('state', $argument);
+                    $data = \GuzzleHttp\json_decode($argument['state'], true);
+                    self::assertArrayHasKey('key', $data[StateFile::NAMESPACE_PREFIX]);
+                    self::assertEquals('fooBar', $data[StateFile::NAMESPACE_PREFIX]['key']);
+                    self::assertArrayHasKey('#foo', $data[StateFile::NAMESPACE_PREFIX]);
+                    self::assertStringStartsWith('KBC::ProjectSecure::', $data[StateFile::NAMESPACE_PREFIX]['#foo']);
+                    return true;
+                })
             );
 
         /** @var Client $sapiStub */
         $stateFile = new StateFile(
             $this->dataDir,
             $sapiStub,
             $this->encryptorFactory,
-            [],
+            ['key' => 'fooBarBaz'],
             'json',
             'docker-demo',
             'config-id',
             new NullFilter()
         );
-        $stateFile->stashState(['state' => 'fooBar']);
+        $stateFile->stashState(["key" => "fooBar", "#foo" => "bar"]);
         $stateFile->persistState();
     }
 
-    public function tesChangeToEmptyArray()
+    public function testStashStateDoesNotUpdate()
+    {
```

Comment:
notetoself: bude se to chovat tak, ze i kdyz se state nezmeni, tak se po behu komponenty zanori, coz je asi spravne

Reply:
hm, tohle bych ještě ověřil

Reply:
udělal jsem na to testík `testExecutorStoreStateDoesNotMigrateUnchangedState` :-)

state, který se nezmění, se nezmigruje.

Reply:
no ono je nezmeni a nezmeni! jedna nezmena je tady https://github.com/keboola/docker-bundle/blob/najlos-state-namespaces/src/Docker/Runner/StateFile.php#L115 a druha nezmena je tady https://github.com/keboola/docker-bundle/blob/najlos-state-namespaces/src/Docker/Runner/StateFile.php#L125

Reply:
to je ale stejne divny, ten test by mel failovat

Reply:
tahle podmínka vrátí false a state se neupdatuje

https://github.com/keboola/docker-bundle/blob/611849264b27e061983f36e20a00548637580c48/src/Docker/Runner/StateFile.php#L125

figl je v tom, že previous state není to, co je uložený u konfigurace, ale už ze zanořenýho `component` nodu, pokud je to namespacovanej state

Reply:
no prave - previous stat neni to co je ulozny u konfigurace a proto to ma vratit true, ne?

Reply:
tak bacha, pokud je u konfigurace uložený

```
# konfigurace
state: {key: val}

# StateFile
$previousState === {key: val}
```

```
# konfigurace
state: {component: {key: val}}}

# StateFile
$previousState === {key: val}
```

protože https://github.com/keboola/docker-bundle/blob/611849264b27e061983f36e20a00548637580c48/src/Docker/Runner/StateFile.php#L88-L92

Reply:
ajo, on uz je vynorenej! jo, makes sense.

i kdyz vlastne premyslim jestli tam je ta podminka ` if (serialize($this->currentState) != serialize($previousState)) { ` vubec k necemu dobra

Reply:
napsal bys to nějak srozumitelnějc?

jinak mi přijde zbytečný ukládat něco, co už uložený je.

Reply:
> jinak mi přijde zbytečný ukládat něco, co už uložený je.

premature optimization... 

je to zajimava otazka v souvislosti s verzovanim - pokud by se nekdy state verzoval, chteli bysme, aby se tam ukladala informace, ze komponenta vyprodukovala stav, ktery je (mozna shodou okolnosti?) stejny jako ulozeny stav? srov. s touhle podminkou https://github.com/keboola/docker-bundle/blob/611849264b27e061983f36e20a00548637580c48/src/Docker/Runner/StateFile.php#L115 ktera rika, ze pokud komponenta nevyprodukovala zadny novy stav, tak se tam necha ten puvodni.

Reply:
stejně jako u konfigurací - pokud uložíš totéž, co tam teď je, tak se nová verze neudělá. furt nerozumím, co je na té podmínce špatně, doteď fungovala, ničemu nevadila, tak proč jí odstranit? test žádnej nezmizí.

Reply:
ja nerikam ze je spatne, nebo ze necemu vadi, ale je tam pravdepodobne uplne zbytecna

a to "pravdepodobne" me drazdi

Reply:
oukej!